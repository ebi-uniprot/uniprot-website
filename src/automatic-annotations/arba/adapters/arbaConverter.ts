import { AAModel } from '../../shared/model';

export type ARBAAPIModel = AAModel & {
  // TODO: change that in the model (backend)
  uniRuleId: string;
};

export type ARBAUIModel = ARBAAPIModel & {
  // any addition/change by the converter
};

const uniRuleConverter = (data: ARBAAPIModel): ARBAUIModel => ({
  ...data,
});

export default uniRuleConverter;
