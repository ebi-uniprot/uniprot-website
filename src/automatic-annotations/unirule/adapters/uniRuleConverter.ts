import { AAModel } from '../../shared/model';

export type UniRuleAPIModel = AAModel & {
  uniRuleId: string;
};

export type UniRuleUIModel = UniRuleAPIModel & {
  // any addition/change by the converter
};

const uniRuleConverter = (data: UniRuleAPIModel): UniRuleUIModel => ({
  ...data,
});

export default uniRuleConverter;
