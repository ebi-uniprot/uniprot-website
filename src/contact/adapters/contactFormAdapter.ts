import { AxiosRequestConfig } from 'axios';
import { debounce } from 'lodash-es';
import {
  Dispatch,
  FormEventHandler,
  SetStateAction,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { useLocation, useNavigate } from 'react-router';

import { Location, LocationToPath } from '../../app/config/urls';
import { addMessage } from '../../messages/state/messagesActions';
import {
  MessageFormat,
  MessageLevel,
} from '../../messages/types/messagesTypes';
import useDataApi from '../../shared/hooks/useDataApi';
import useMessagesDispatch from '../../shared/hooks/useMessagesDispatch';
import { stringifyUrl } from '../../shared/utils/url';
import apiUrls from '../config/apiUrls';

export const modifyFormData = (formData: FormData, token: string) => {
  const output = new FormData();
  output.set('token', token);
  output.set('email', formData.get('email') || '');
  output.set('subject', `[uuw] ${formData.get('subject') || ''}`);
  output.set('requiredForRobots', formData.get('requiredForRobots') || '');
  const message = `${formData.get('message') || ''}

Name: ${formData.get('name')}

--------------- prefilled context details ---------------

${formData.get('context') || ''}`;
  output.set('message', message);
  return output;
};

const DEBOUNCE_TIME = 1_000;

export type Suggestion =
  | 'update'
  | 'PDB'
  | 'buy'
  | 'blast'
  | 'align'
  | 'id mapping'
  | 'peptide search'
  | 'not English';
const entryUpdateRegExp = /entry update request/i;
const pdbRegExp = /(structure)|(pdb)/i;
const buyRegExp = /(buy)|(purchas)|(ship)|(cost)|(quote)/i;
const blastRegExp = /blast/i;
const alignRegExp = /(clustal)|(align)/i;
const idMappingRegExp = /(map)/i;
const peptideSearchRegExp = /(peptide search)|(search peptide)/i;
// anything not outside the printable ASCII range
// eslint-disable-next-line no-control-regex
let nonLatin = /[^\x00-\x7F]/;
try {
  // any Unicode letter minus characters whose Script Extensions include Latin
  // α: true, 東: true, t: false, é: false
  nonLatin = /[\p{L}--\p{scx=Latin}]/v;
} catch {
  /* */
}

export const getSuggestion = debounce(
  (
    callback: Dispatch<SetStateAction<Suggestion[] | undefined>>,
    subject: string = '',
    message: string = ''
  ): void => {
    const suggestions: Suggestion[] = [];
    if (entryUpdateRegExp.test(subject)) {
      // A bit specific, only test the subject as it would have been added by
      // the link from entry page to suggest an update
      suggestions.push('update');
    }
    if (pdbRegExp.test(subject) || pdbRegExp.test(message)) {
      suggestions.push('PDB');
    }
    if (buyRegExp.test(subject) || buyRegExp.test(message)) {
      suggestions.push('buy');
    }
    if (blastRegExp.test(subject) || blastRegExp.test(message)) {
      suggestions.push('blast');
    }
    if (alignRegExp.test(subject) || alignRegExp.test(message)) {
      suggestions.push('align');
    }
    if (idMappingRegExp.test(subject) || idMappingRegExp.test(message)) {
      suggestions.push('id mapping');
    }
    if (
      peptideSearchRegExp.test(subject) ||
      peptideSearchRegExp.test(message)
    ) {
      suggestions.push('peptide search');
    }
    if (nonLatin.test(message)) {
      suggestions.push('not English');
    }
    if (!suggestions.length) {
      callback(undefined);
    } else {
      callback(suggestions);
    }
  },
  DEBOUNCE_TIME
);

export type UseFormLogicReturnType = {
  sending: boolean;
  suggestions: undefined | Suggestion[];
  handleSubmit: FormEventHandler<HTMLFormElement>;
  handleChange: FormEventHandler<HTMLInputElement | HTMLTextAreaElement>;
};

export const useFormLogic = (referrer?: string): UseFormLogicReturnType => {
  const dispatch = useMessagesDispatch();
  const navigate = useNavigate();
  const { state } = useLocation();

  const [formData, setFormData] = useState<undefined | FormData>();
  const [suggestions, setSuggestions] = useState<undefined | Suggestion[]>(
    undefined
  );

  const subject = formData?.get('subject') as undefined | string;

  const tokenData = useDataApi<{ token: string }>(
    subject ? stringifyUrl(apiUrls.token, { key: subject }) : null
  );
  const token = tokenData.data?.token;

  // Generate configuration to POST data to the contact send endpoint
  const postData: undefined | AxiosRequestConfig<FormData> = useMemo(() => {
    if (!formData || !token) {
      return;
    }
    const modifiedFormData = modifyFormData(formData, token);

    return {
      method: 'POST',
      data: modifiedFormData,
    };
  }, [formData, token]);

  const sendData = useDataApi(postData ? apiUrls.send : null, postData);

  const loading = tokenData.loading || sendData.loading;
  const error = tokenData.error || sendData.error;

  // Error behaviour
  useEffect(() => {
    if (error) {
      // Pop up message for the user
      dispatch(
        addMessage({
          format: MessageFormat.POP_UP,
          level: MessageLevel.FAILURE,
          content: 'Error while sending your message. Please try again later.',
        })
      );
    }
  }, [error, dispatch]);

  // Success behaviour
  useEffect(() => {
    if (sendData.data) {
      // Pop up message for the user
      dispatch(
        addMessage({
          format: MessageFormat.POP_UP,
          level: MessageLevel.SUCCESS,
          content: 'Your message has been succesfully sent to our team.',
        })
      );
      // Clear the current form (just in case)
      document.querySelector<HTMLFormElement>('main form')?.reset();
      // Navigate the user to previous page, or to the homepage if not possible
      navigate(referrer || LocationToPath[Location.Home]);
    }
  }, [sendData.data, dispatch, navigate, referrer]);

  const previousLoading = useRef(loading);
  // Reset custom hook form data (to retry sending if needed)
  useEffect(() => {
    if (previousLoading.current && !loading) {
      setFormData(undefined);
    }
    previousLoading.current = loading;
  }, [loading]);

  const handleSubmit: FormEventHandler<HTMLFormElement> = useCallback(
    (event) => {
      const form = event.target;
      event.preventDefault();
      if (!(form instanceof HTMLFormElement) || loading) {
        return; // Shouldn't happen, but just in case
      }
      setFormData(new FormData(form));
    },
    [loading]
  );

  // on unmount, cancel any pending suggestion calculation
  useEffect(() => getSuggestion.cancel, []);

  const handleChange: FormEventHandler<
    HTMLInputElement | HTMLTextAreaElement
  > = (event) => {
    const element = event.target;
    if (
      !(
        element instanceof HTMLInputElement ||
        element instanceof HTMLTextAreaElement
      ) ||
      loading
    ) {
      return;
    }
    getSuggestion(
      setSuggestions,
      element.form?.subject?.value,
      element.form?.message?.value
    );
    navigate('.', {
      replace: true,
      state: {
        ...state,
        formValues: {
          ...(state?.formValues || {}),
          [element.name]: element.value,
        },
      },
    });
  };

  return { sending: loading, suggestions, handleSubmit, handleChange };
};
