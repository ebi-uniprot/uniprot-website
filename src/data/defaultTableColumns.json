["accession", "id", "protein_name", "gene_names", "organism"]
