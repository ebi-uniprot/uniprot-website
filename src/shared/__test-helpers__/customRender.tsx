/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
