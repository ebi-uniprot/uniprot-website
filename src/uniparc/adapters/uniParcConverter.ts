import { Sequence } from '../../shared/components/entry/SequenceSection';

export type XRefProperty = {
  key: string; // should replace with union of possibilities?
  value: string;
};

// TODO: to be reviewed when working on UniParc pages
export type UniParcXRef = {
  active: boolean;
  created: string;
  database: string; // should replace with union of possibilities?
  id: string;
  lastUpdated: string;
  properties: XRefProperty[];
  version: number;
  versionI: number; // ðŸ¤” what is this?
};

export type SequenceFeatureLocation = {
  start: number;
  end: number;
};

export type SequenceFeature = {
  database: string; // should replace with union of possibilities?
  databaseId: string;
  interproGroup: {
    id: string;
    name: string;
  };
  locations: SequenceFeatureLocation[];
};

export type UniParcAPIModel = {
  uniParcId: string;
  uniParcCrossReferences: UniParcXRef[];
  taxonomies: { taxonId: number }[];
  sequenceFeatures: SequenceFeature[];
  sequence: Sequence;
};

export type UniParcUIModel = UniParcAPIModel & {
  // any addition/change by the converter
};

const uniParcConverter = (data: UniParcAPIModel): UniParcUIModel => ({
  ...data,
});

export default uniParcConverter;
