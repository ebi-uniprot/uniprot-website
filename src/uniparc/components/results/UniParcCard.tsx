import { useMemo } from 'react';
import { Card, LongNumber } from 'franklin-sites';
import { Link } from 'react-router-dom';

import EntryTitle from '../../../shared/components/entry/EntryTitle';
import BasketStatus from '../../../basket/BasketStatus';
import CardCheckboxCell from '../../../shared/components/CardCheckboxCell';

import {
  UniParcAPIModel,
  XRefsInternalDatabasesEnum,
} from '../../adapters/uniParcConverter';

import { getEntryPath } from '../../../app/config/urls';
import { getIdKeyForNamespace } from '../../../shared/utils/getIdKeyForNamespace';
import xrefGetter from '../../utils/xrefGetter';
import { pluralise } from '../../../shared/utils/utils';

import UniParcColumnConfiguration, {
  UniParcColumn,
} from '../../config/UniParcColumnConfiguration';
import RenderColumnsInCard from '../../../shared/components/results/RenderColumnsInCard';

import { Namespace } from '../../../shared/types/namespaces';
import { EntryType } from '../../../shared/components/entry/EntryTypeIcon';

import renderColumnsInCardStyles from '../../../shared/components/results/styles/render-columns-in-card.module.scss';

const mainInfoColumns = [
  UniParcColumn.firstSeen,
  UniParcColumn.lastSeen,
  UniParcColumn.length,
].map((column) => UniParcColumnConfiguration.get(column));

const uniProtKBCounter = (data: UniParcAPIModel) => {
  let reviewed = 0;
  let unreviewed = 0;
  for (const xref of data.uniParcCrossReferences || []) {
    if (xref.database === XRefsInternalDatabasesEnum.REVIEWED) {
      reviewed += 1;
    } else if (xref.database === XRefsInternalDatabasesEnum.UNREVIEWED) {
      unreviewed += 1;
    }
  }
  return { reviewed, unreviewed };
};

const getIdKey = getIdKeyForNamespace(Namespace.uniparc);

const UniParcCard = ({ data }: { data: UniParcAPIModel }) => {
  const id = getIdKey(data);

  const organismCount = xrefGetter(data, 'organism', 'taxonId')?.length || 0;
  const uniProtKBCount = useMemo(() => uniProtKBCounter(data), [data]);

  return (
    <Card
      header={
        <>
          <CardCheckboxCell id={id} />
          <h2 className="small">
            <Link to={getEntryPath(Namespace.uniparc, id)}>
              <EntryTitle mainTitle={id} entryType={EntryType.UNIPARC} />
            </Link>
          </h2>
          <BasketStatus id={id} className="tiny" />
        </>
      }
      headerSeparator={false}
    >
      <div className={renderColumnsInCardStyles['result-card__info-container']}>
        <span className={renderColumnsInCardStyles['result-card__info-bit']}>
          <strong>{pluralise('Organism', organismCount)}: </strong>
          <LongNumber>{organismCount}</LongNumber>
        </span>
        <span className={renderColumnsInCardStyles['result-card__info-bit']}>
          <strong>UniprotKB entries: </strong>
          <LongNumber>{uniProtKBCount.reviewed}</LongNumber> reviewed and{' '}
          <LongNumber>{uniProtKBCount.unreviewed}</LongNumber> unreviewed
        </span>
      </div>
      <RenderColumnsInCard data={data} renderers={mainInfoColumns} />
    </Card>
  );
};

export default UniParcCard;
