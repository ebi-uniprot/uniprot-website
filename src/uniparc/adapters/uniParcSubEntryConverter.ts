import { getSource, getXRefsForId, isSourceDatabase } from '../utils/subEntry';
import uniParcConverter, {
  UniParcLiteAPIModel,
  UniParcUIModel,
  UniParcXRef,
  databaseToEntryType,
} from './uniParcConverter';

export type UniParcSubEntryUIModel = {
  entry: Partial<UniParcUIModel>;
  subEntry: Partial<UniParcXRef> & {
    isSource: boolean;
    source?: Partial<UniParcXRef> | null;
    isUniprotkbEntry: boolean;
  };
};

const uniParcSubEntryConverter = (
  entryData: UniParcLiteAPIModel,
  subEntryId: string
): UniParcSubEntryUIModel | null => {
  const transformedEntryData = uniParcConverter(entryData);

  const subEntryData = getXRefsForId(subEntryId);

  if (!subEntryData) {
    return null;
  }
  const isUniprotkbEntry = Boolean(
    subEntryData.database && databaseToEntryType.has(subEntryData.database)
  );

  const isSource = isSourceDatabase(subEntryData.database);
  const source = isSource ? undefined : getSource(subEntryData.organism);

  return {
    entry: transformedEntryData,
    subEntry: { ...subEntryData, isSource, source, isUniprotkbEntry },
  };
};

export default uniParcSubEntryConverter;
