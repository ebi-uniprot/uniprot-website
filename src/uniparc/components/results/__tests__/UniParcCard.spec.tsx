import { fireEvent, screen } from '@testing-library/react';

import customRender from '../../../../shared/__test-helpers__/customRender';

import UniParcCard from '../UniParcCard';

import { UniParcAPIModel } from '../../../adapters/uniParcConverter';

import data from '../../../__mocks__/entryModelData';

describe('UniRefCard tests', () => {
  it('should render and match snapshot', () => {
    const { asFragment } = customRender(
      <UniParcCard
        data={data as UniParcAPIModel}
        handleEntrySelection={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should allow card selection and navigation', () => {
    const handleClick = jest.fn();
    const { history } = customRender(
      <UniParcCard
        data={data as UniParcAPIModel}
        handleEntrySelection={handleClick}
      />
    );
    fireEvent.click(screen.getByRole('checkbox'));
    expect(handleClick).toHaveBeenCalled();
    expect(history.location.pathname).not.toMatch('/uniparc/UPI0000000001');
    fireEvent.click(screen.getByTestId('background-link'));
    expect(history.location.pathname).toMatch('/uniparc/UPI0000000001');
  });
});
