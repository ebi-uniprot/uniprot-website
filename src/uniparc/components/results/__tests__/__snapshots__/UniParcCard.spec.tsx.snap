// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniRefCard tests should render and match snapshot 1`] = `
<DocumentFragment>
  <section
    class="card"
  >
    <div
      class="card__container"
    >
      <div
        class="card__header"
      >
        <span
          class="checkbox-cell"
        >
          <input
            data-id="UPI0000000001"
            id="1"
            type="checkbox"
          />
          <label
            aria-label="UPI0000000001"
            for="1"
            title="click to select, shift+click for multiple selection"
          />
        </span>
        <h2
          class="small"
        >
          <a
            href="/uniparc/UPI0000000001"
          >
            <span
              class="entry-title"
              translate="no"
            >
              <span
                class="entry-title__status icon--uniparc"
                title="This marks a UniParc entry"
              >
                <test-file-stub />
              </span>
              UPI0000000001
            </span>
          </a>
        </h2>
      </div>
      <div
        class="card__content"
      >
        <div
          class="result-card__info-container"
        >
          <span
            class="result-card__info-bit"
          >
            <strong>
              Organisms: 
            </strong>
            7
          </span>
          <span
            class="result-card__info-bit"
          >
            <strong>
              UniprotKB entries: 
            </strong>
            2 reviewed and 7 unreviewed
          </span>
        </div>
        <div
          class="result-card__info-container"
        >
          <span
            class="result-card__info-bit"
          >
            <strong>
              <span>
                First seen
              </span>
              : 
            </strong>
            <time
              datetime="1988-11-01T00:00:00.000Z"
            >
              1988-11-01
            </time>
          </span>
          <span
            class="result-card__info-bit"
          >
            <strong>
              <span>
                Last seen
              </span>
              : 
            </strong>
            <time
              datetime="2024-07-24T00:00:00.000Z"
            >
              2024-07-24
            </time>
          </span>
          <span
            class="result-card__info-bit"
          >
            <strong>
              <span>
                Length
              </span>
              : 
            </strong>
            250
          </span>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
